mlcomp/parallelm/pipeline/dag.py

```diff
@@ -129,32 +128,44 @@ def sorted_execution_graph(self):
 
     def parent_data_objs(self, dag_node):
         data_objs = []
+        # dict to keep track of proper input index going to components
+        data_object_dict = {}
+        max_index = 0
+
         for parent in dag_node.parents():
             parent_id = parent[json_fields.PIPELINE_COMP_PARENTS_FIRST_FIELD]
             output_index = parent[json_fields.PIPELINE_COMP_PARENTS_SECOND_FIELD]
+
+            input_index = parent.get(json_fields.PIPELINE_COMP_PARENTS_THIRD_FIELD, max_index)
+
             if parent_id in self._parent_data_objs_placeholder:
                 if output_index in self._parent_data_objs_placeholder[parent_id]:
                     self._logger.debug("Concatenate parent data objs, parent_id={}, output_index={}"
                                        .format(parent_id, output_index))
-                    data_objs.append(self._parent_data_objs_placeholder[parent_id][output_index])
+                    data_object_dict[input_index] = self._parent_data_objs_placeholder[parent_id][output_index]
                 else:
-                    data_objs.append(None)
+                    data_object_dict[input_index] = None
                     self._logger.debug("Output index not in data objs placeholder! parent_id={}, output_index={}"
                                        .format(parent_id, output_index))
             else:
-                data_objs.append(None)
+                data_object_dict[input_index] = None
                 self._logger.debug("Parent id not in data objs placeholder! id={}".format(parent_id))
+
+            max_index += 1
+
+        for index in range(max_index):
```

Comment:
just to be safe (1) ensure that in fact, we cover all values from 0..max_index when we setup the dict (this may have already been verified by dag gen, but assert it here)
(2) max_index == len(dag_node.parents())? we can skip that variable i guess?

Reply:
max_index is used in getting parent's input index. if not specified then it will be used as per max_index.

Reply:
assertion is added though! but keeping that variable alive for enow